### PR TITLE
Replace `int` with `SupportsIndex` in indexing methods hints

### PR DESCRIPTION
**Pull request by honno**
Resolves https://github.com/data-apis/array-api/issues/383

### CODE REVIEW

File: src/array_api_stubs/_draft/array_object.py [hunk 1]
```diff
@@ -1089,11 +1098,13 @@ def __setitem__(
...
         self: array
             array instance.
-        key: Union[int, slice, ellipsis, Tuple[Union[int, slice, ellipsis], ...], array]
+        key: Union[SupportsIndex, slice, ellipsis, Tuple[Union[SupportsIndex, slice, ellipsis], ...], array]
             index key.
         value: Union[int, float, bool, array]
             value(s) to set. Must be compatible with ``self[key]`` (see :ref:`broadcasting`).
 
+        .. note::
+           ``key`` can only be an array if it is valid for boolean array indexing, or supports ``__index__()``.
```
Comment (honno) [inline src/array_api_stubs/_draft/array_object.py:1107]: I really can't figure out why this gives a warning—any playing around with notes for `__setitem__()` gives me the same thing  :thinking:

```
/home/honno/PycharmProjects/array-api/src/array_api_stubs/_draft/array_object.py:docstring of array_api_stubs._draft.array_object._array.__setitem__:1: WARNING: py:class reference target not found: ..
```

**Comment by rgommers** [comment]
Thanks @honno.

Adding a quick clarifying example:
```python
>>> import numpy as np
>>> import torch
>>> 
>>> x = np.arange(1, 4)
>>> y = torch.arange(1, 4)
>>> 
>>> class Ix:
...     def __index__(self):
...         return 1
... 
>>> x[Ix()]
2
>>> y[Ix()]
tensor(2)
```

File: src/array_api_stubs/_draft/array_object.py (hunk 1 above)
Comment (rgommers) [reply src/array_api_stubs/_draft/array_object.py:1107]: There is `nitpick_ignore` list in `src/_array_api_conf.py` that may be needed here.

Comment (honno) [reply src/array_api_stubs/_draft/array_object.py:1109]: Can ignore the re-review-request still need to figure this one out :sweat_smile:

**Comment by rgommers** [comment]
@honno would you be able to open a PR with a test for this to `array-api-tests`? Doesn't have to be merged before this will be merged, but that will help smoke out if any known/tested library does not yet support this feature.

File: src/array_api_stubs/_draft/array_object.py [hunk 2]
```diff
@@ -620,9 +622,13 @@ def __getitem__(
         ----------
         self: array
             array instance.
-        key: Union[int, slice, ellipsis, None, Tuple[Union[int, slice, ellipsis, None], ...], array]
+        key: Union[int, slice, ellipsis, None, SupportsIndex, Tuple[Union[int, slice, ellipsis, None, SupportsIndex], ...], array]
             index key.
 
+
+        .. note::
+           ``key`` can only be an array if it is valid for boolean array indexing, or supports ``__index__()``.
```
Comment (asmeurer) [inline src/array_api_stubs/_draft/array_object.py:630]: Presumably somewhere else we say it should support `__index__` if and only if it is a 0-D integer array. Maybe it would be clearer to just say that directly here.

**Comment by honno** [comment]
> @honno would you be able to open a PR with a test for this to `array-api-tests`? Doesn't have to be merged before this will be merged, but that will help smoke out if any known/tested library does not yet support this feature.

Good shout, I opened https://github.com/data-apis/array-api-tests/pull/247 to check this all out. From the looks of it:

* NumPy (+ `array_api_strict`), PyTorch
* JAX doesn't support, has a general note to open a feature requests on indexing modes (... that's prob refering to advance integer indexing).
* cupy seems to support advance integer indexing, but doesn't play nice with 0-D integer arrays as indexes.arrays used as indices must be of integer (or boolean) type
* Dask at least accepts 0-D integer input, but `array_api_compat.dask` not playing nice with the test suite so I'll have to triage that and explore further.

File: src/array_api_stubs/_draft/array_object.py (hunk 2 above)
Comment (honno) [reply src/array_api_stubs/_draft/array_object.py]: Yep that'd help, let me know on the wording from my latest commit.

**Comment by asmeurer** [comment]
> but array_api_compat.dask not playing nice with the test suite so I'll have to triage that and explore further.

Let me know what you find out. We do run it on CI with some [skips](https://github.com/data-apis/array-api-compat/blob/main/dask-skips.txt) and [xfails](https://github.com/data-apis/array-api-compat/blob/main/dask-xfails.txt), and also the [max-examples is set to 5](https://github.com/data-apis/array-api-compat/blob/main/.github/workflows/array-api-tests-dask.yml). I haven't looked at the Dask xfails too closely, and obviously if we can remove any of those that would be great. CC @lithomas1

**Comment by kgryte** [comment]
@honno Were you able to triage the Dask issues with the test suite?

**Comment by asmeurer** [comment]
By the way, for implementers, the generally correct behavior is to `operator.index()` to normalize index objects that aren't one of the other supported index types like Ellipsis, slice, or array. My guess is that dask.array isn't doing that.

**Comment by kgryte** [comment]
@honno Is there anything more that we need to do with this PR?

**Comment by honno** [comment]
> @honno Is there anything more that we need to do with this PR?

PR I think I'm happy with, just be mindful that last I checked in March it was only NumPy and PyTorch that supported "indexables", whereas JAX/CuPy/Dask didn't, so they'd need updating to support this.

Example of what I mean by an indexable:

```python
class AwkwardIndexable:
    def __init__(self, value: int):
        self._value = value

    def __int__(self):
        raise TypeError("__int__() should not be called")

    def __index__(self):
        return self._value
``
